bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Cover.java

```diff
@@ -48,18 +57,97 @@ static class ChannelConfiguration extends AbstractChannelConfiguration {
         protected String payloadClose = "CLOSE";
         @SerializedName("payload_stop")
         protected String payloadStop = "STOP";
+        @SerializedName("position_closed")
+        protected int positionClosed = 0;
+        @SerializedName("position_open")
+        protected int positionOpen = 100;
+        @SerializedName("position_template")
+        protected @Nullable String positionTemplate;
+        @SerializedName("position_topic")
+        protected @Nullable String positionTopic;
+        @SerializedName("set_position_template")
+        protected @Nullable String setPositionTemplate;
+        @SerializedName("set_position_topic")
+        protected @Nullable String setPositionTopic;
+        @SerializedName("state_closed")
+        protected String stateClosed = "closed";
+        @SerializedName("state_closing")
+        protected String stateClosing = "closing";
+        @SerializedName("state_open")
+        protected String stateOpen = "open";
+        @SerializedName("state_opening")
+        protected String stateOpening = "opening";
+        @SerializedName("state_stopped")
+        protected String stateStopped = "stopped";
     }
 
+    @Nullable
+    ComponentChannel stateChannel = null;
+
     public Cover(ComponentFactory.ComponentConfiguration componentConfiguration) {
         super(componentConfiguration, ChannelConfiguration.class);
 
-        RollershutterValue value = new RollershutterValue(channelConfiguration.payloadOpen,
-                channelConfiguration.payloadClose, channelConfiguration.payloadStop);
+        String stateTopic = channelConfiguration.stateTopic;
+
+        // State can indicate additional information than just
+        // the current position, so expose it as a separate channel
+        if (stateTopic != null) {
+            TextValue value = new TextValue(new String[] { channelConfiguration.stateClosed,
+                    channelConfiguration.stateClosing, channelConfiguration.stateOpen,
+                    channelConfiguration.stateOpening, channelConfiguration.stateStopped });
+            buildChannel(STATE_CHANNEL_ID, value, "State", componentConfiguration.getUpdateListener())
+                    .stateTopic(stateTopic).isAdvanced(true).build();
+        }
+
+        if (channelConfiguration.commandTopic != null) {
+            hiddenChannels.add(stateChannel = buildChannel(STATE_CHANNEL_ID, new TextValue(), "State",
```

Comment:
What will happen if stateChannel is null ?

Reply:
I'm not sure which part you're asking about. In this line, it's _assigning_ to stateChannel, and it can't be null. Down further in the file, when the rollershutter channel receives UP, DOWN, or STOP (instead of a PercentType), if the state channel is not null, it sends the command to the stateChannel, rather than the rollershutter channel (which is linked to a command topic that can only accept numbers).

Reply:
Sorry, I read + instead of =